onnxruntime/core/providers/cpu/tensor/gather.cc

```diff
@@ -49,11 +56,14 @@ Status GatherCopyData(const Tensor* indices_tensor, const uint8_t* src_base, uin
 
   // Check the indices first in case there's a out of bound index.
   // We can't merge this code in the omp loop below as omp does not allow return in the loop
+  auto axis_dim_limit = input_data_shape[axis];
+
   for (int64_t i = 0; i < N; ++i) {
     Tin idx = indices_data[i];
-    if (idx < 0 || idx >= input_data_shape[axis]) {
-      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "indices element out of data bounds, idx=", idx,
-                             " data_dim=", input_data_shape[axis]);
+    if (idx < -axis_dim_limit || idx >= axis_dim_limit) {
```

Comment:
We have internal discussion about supporting this for opset-10. Keeping this comment until we have a conclusion.